asr1k_neutron_l3/extensions/vpn_asr1k_int_tunnel_ips.py

```
@@ -0,0 +1,94 @@
+# Copyright 2025 SAP SE
+#
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from neutron_lib.api import converters
+from neutron_lib.api.definitions import vpn
+from neutron_lib.api import extensions as api_extensions
+from neutron_lib import constants
+from neutron_lib.db import resource_extend
+
+
+class ApiDef:
+    ALIAS = 'vpn-asr1k-int-tunnel-ips'
+    IS_SHIM_EXTENSION = False
+    IS_STANDARD_ATTR_EXTENSION = False
+    NAME = 'VPNaaS Internal Tunnel IP Addresses (asr1k flavor)'
+    DESCRIPTION = "Allow giving the tunnel an internal v4 and v6 prefix"
+    UPDATED_TIMESTAMP = "2025-09-22T08:00:00-00:00"
+    API_PREFIX = '/vpn'
+
+    RESOURCE_ATTRIBUTE_MAP = {
+        vpn.IPSEC_SITE_CONNECTIONS: {
+            # internal cidr v4
+            'int_local_cidr_v4': {
+                'allow_post': True, 'allow_put': True,
+                'default': constants.ATTR_NOT_SPECIFIED,
+                'convert_to': converters.convert_cidr_to_canonical_format,
+                'validate': {'type:subnet_or_none': None},
```

do we need to make sure that only addresses/subnets with the correct IP version are supplied? Could something break later / when configuring?

---

Yes, but I haven't seen neutron having an extra validator for that. We should handle this in `vpnaas_driver.py` in `validate_internal_tunnel_ips()`

---

We could also implement a validator and try to get that upstream later, it just feels like a lack of input validation here and does not follow the 'fail early' -- but it is what it is: an existing code base that we need to adapt to. Would add a comment in the source though that this does not validate the IP version and we are doing that in the other method to document that decision/behavior.

---

There is a lot of validation that we do afterwards, so I don't know if a comment that says "this field needs further validation" is needed. In my mind it's "just" the API def with some preliminary validation and rest is done elsewhere.

I've written a test and some validation code that I'll push up with the first force push.